File: programmers/258705.cpp
#include <string>
#include <vector>

using namespace std;

#define N 101010
#define MOD 10007

int solution(int n, vector<int> tops) {
    int answer;
    int d[N][2];
    
    if (tops[0]) {
        d[0][0] = 3;
        d[0][1] = 1;
    }
    else {
        d[0][0] = 2;
        d[0][1] = 1;
    }
    for (int i = 1; i < n; i++) {
        if (tops[i]) {
            d[i][0] = (d[i - 1][0] * 3 + d[i - 1][1] * 2) % MOD;
            d[i][1] = (d[i - 1][0] + d[i - 1][1]) % MOD;
        }
        else {
            d[i][0] = (d[i - 1][0] * 2 + d[i - 1][1]) % MOD;
            d[i][1] = (d[i - 1][0] + d[i - 1][1]) % MOD;
        }
    }
    answer = (d[n - 1][0] + d[n - 1][1]) % MOD;
    
    return answer;
}

File: programmers/258705.java
class Solution {
    private static final int MOD = 10007;
    
    public int solution(int n, int[] tops) {
        int[][] dp = new int[n][4];
        
        dp[0][0] = 1;                       // 정삼각형 타일로만 채우는 경우
        dp[0][1] = 1;                       // 2-8시 방향 마름모 타일을 놓는 경우
        dp[0][2] = 1;                       // 4-10시 방향 마름모 타일을 놓는 경우
        dp[0][3] = (tops[0] == 1) ? 1 : 0;  // 6-12시 방향 마름모 타일을 놓는 경우
        for (int i = 1; i < n; i++) {
            dp[i][0] = (dp[i - 1][0] + dp[i - 1][1] + dp[i - 1][2] + dp[i - 1][3]) % MOD;
            dp[i][1] = (dp[i - 1][0] + dp[i - 1][1] + dp[i - 1][3]) % MOD;
            dp[i][2] = dp[i][0];
            dp[i][3] = (tops[i] == 1) ? dp[i][0] : 0; 
        }
        
        return (dp[n - 1][0] + dp[n - 1][1] + dp[n - 1][2] + dp[n - 1][3]) % MOD;
    }
}